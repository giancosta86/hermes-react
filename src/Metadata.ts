import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation?: string;
  className?: string;
}>;

export type CharMetadataMap = RMap<Char, CharMetadata>;

export type MetadataByChar = Readonly<{ [char: Char]: CharMetadata }>;
