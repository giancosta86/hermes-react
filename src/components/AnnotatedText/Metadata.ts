import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation: string;
  classes: Iterable<string>;
}>;

export type MetadataByChar = RMap<Char, CharMetadata>;
