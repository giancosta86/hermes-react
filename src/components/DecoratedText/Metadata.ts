import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation?: string;
  className?: string;
}>;

export type MetadataByChar = RMap<Char, CharMetadata>;
