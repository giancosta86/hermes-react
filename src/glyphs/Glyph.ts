export namespace Glyph {
  export const latin = "latin";
  export const nonLatin = "non-latin";
  export const phonogram = "phonogram";
  export const logogram = "logogram";
  export const chinese = "chinese";
  export const pinyin = "pinyin";
}
