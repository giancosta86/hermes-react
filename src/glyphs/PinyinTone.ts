export const flat = "flat";

export const rising = "rising";

export const fallingRising = "falling-rising";

export const falling = "falling";

export const neuter = "neuter";
