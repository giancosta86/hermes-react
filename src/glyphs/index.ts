export * as Glyph from "./Glyph";

export * as Pinyin from "./Pinyin";

export * as PinyinTone from "./PinyinTone";

export * as Russian from "./russian";
