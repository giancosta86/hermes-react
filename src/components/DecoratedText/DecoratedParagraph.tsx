import React, { useMemo } from "react";
import { MetadataByChar } from "./Metadata";
import { DecoratedChar } from "./DecoratedChar";

export interface DecoratedParagraphProps {
  text: string;
  metadataByChar: MetadataByChar;
}

export const DecoratedParagraph = ({
  text,
  metadataByChar
}: DecoratedParagraphProps) => {
  const chars = useMemo(() => text.split(""), [text]);

  return (
    <>
      {chars.map((char, index) => {
        if (char == "\n") {
          return <br key={index} />;
        }

        const metadata = metadataByChar.get(char);

        return metadata && (metadata.annotation || metadata.className) ? (
          <DecoratedChar key={index} char={char} metadata={metadata} />
        ) : (
          <React.Fragment key={index}>{char}</React.Fragment>
        );
      })}
    </>
  );
};
