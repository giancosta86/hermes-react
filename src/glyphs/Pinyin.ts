import classnames from "classnames";
import { Glyph } from "./Glyph";

export namespace PinyinTone {
  export const flat = "flat";

  export const rising = "rising";

  export const fallingRising = "falling-rising";

  export const falling = "falling";

  export const neuter = "neuter";
}

export namespace Pinyin {
  const baseClasses = [
    Glyph.nonLatin,
    Glyph.logogram,
    Glyph.chinese,
    Glyph.pinyin
  ] as const;

  export const flat = classnames(...baseClasses, PinyinTone.flat);

  export const rising = classnames(...baseClasses, PinyinTone.rising);

  export const fallingRising = classnames(
    ...baseClasses,
    PinyinTone.fallingRising
  );

  export const falling = classnames(...baseClasses, PinyinTone.falling);

  export const neuter = classnames(...baseClasses, PinyinTone.neuter);
}
