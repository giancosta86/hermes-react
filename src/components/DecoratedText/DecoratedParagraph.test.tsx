import React from "react";
import { create } from "react-test-renderer";
import { HashMap } from "@rimbu/hashed";
import { DecoratedParagraph } from "./DecoratedParagraph";
import { Char, CharMetadata } from "./Metadata";
import { Pinyin } from "@/glyphs";

describe("<DecoratedParagraph>", () => {
  describe("when receiving a paragraph with no metadata", () => {
    it("should render just plain characters", () => {
      const renderer = create(
        <DecoratedParagraph text="Dodo" metadataByChar={HashMap.empty()} />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with one intermediate newline", () => {
    it("should render the newline as a <br/>", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"Alpha\nBeta"}
          metadataByChar={HashMap.empty()}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with multiple intermediate newlines", () => {
    it("should render every newline as a <br/>", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"Alpha\nBeta\nGamma"}
          metadataByChar={HashMap.empty()}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with one character having annotation and classes", () => {
    it("should render that character with a <ruby> tag having classes", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我爱看书，你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>([
            "书",
            { annotation: "shū", className: Pinyin.flat }
          ])}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with one character having annotation but no classes", () => {
    it("should render that character with a classless <ruby> tag", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我爱看书，你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>([
            "书",
            { annotation: "shū" }
          ])}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with one character having neither annotation nor classes", () => {
    it("should render that character with no tags", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我爱看书，你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>(["书", {}])}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with one character having classes but no annotation", () => {
    it("should render that character with a <span> tag", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我爱看书，你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>([
            "书",
            { className: Pinyin.flat }
          ])}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with multiple characters having annotation and classes", () => {
    it("should render such characters with <ruby> tags having classes", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我爱看书，你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>(
            ["看", { annotation: "kàn", className: Pinyin.falling }],
            ["书", { annotation: "shū", className: Pinyin.flat }],
            ["呢", { annotation: "ne", className: Pinyin.neuter }]
          )}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving a paragraph with multiple characters having annotation and/or classes, as well as newlines", () => {
    it("should render everything in a suitable way", () => {
      const renderer = create(
        <DecoratedParagraph
          text={"我\n爱看书，\n你呢？"}
          metadataByChar={HashMap.of<Char, CharMetadata>(
            ["看", { annotation: "kàn" }],
            ["书", { annotation: "shū", className: Pinyin.flat }],
            ["呢", { className: Pinyin.neuter }]
          )}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });
});
