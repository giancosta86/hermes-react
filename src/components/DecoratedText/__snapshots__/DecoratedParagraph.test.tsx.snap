// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DecoratedParagraph> when receiving a paragraph with multiple characters having annotation and classes should render such characters with <ruby> tags having classes 1`] = `
[
  "我",
  "爱",
  <ruby
    className="non-latin logogram chinese pinyin falling"
  >
    看
    <rp>
      (
    </rp>
    <rt>
      kàn
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  <ruby
    className="non-latin logogram chinese pinyin flat"
  >
    书
    <rp>
      (
    </rp>
    <rt>
      shū
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  "，",
  "你",
  <ruby
    className="non-latin logogram chinese pinyin neuter"
  >
    呢
    <rp>
      (
    </rp>
    <rt>
      ne
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with multiple characters having annotation and/or classes, as well as newlines should render everything in a suitable way 1`] = `
[
  "我",
  <br />,
  "爱",
  <ruby>
    看
    <rp>
      (
    </rp>
    <rt>
      kàn
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  <ruby
    className="non-latin logogram chinese pinyin flat"
  >
    书
    <rp>
      (
    </rp>
    <rt>
      shū
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  "，",
  <br />,
  "你",
  <span
    className="non-latin logogram chinese pinyin neuter"
  >
    呢
  </span>,
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with multiple intermediate newlines should render every newline as a <br/> 1`] = `
[
  "A",
  "l",
  "p",
  "h",
  "a",
  <br />,
  "B",
  "e",
  "t",
  "a",
  <br />,
  "G",
  "a",
  "m",
  "m",
  "a",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with no metadata should render just plain characters 1`] = `
[
  "D",
  "o",
  "d",
  "o",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with one character having annotation and classes should render that character with a <ruby> tag having classes 1`] = `
[
  "我",
  "爱",
  "看",
  <ruby
    className="non-latin logogram chinese pinyin flat"
  >
    书
    <rp>
      (
    </rp>
    <rt>
      shū
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  "，",
  "你",
  "呢",
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with one character having annotation but no classes should render that character with a classless <ruby> tag 1`] = `
[
  "我",
  "爱",
  "看",
  <ruby>
    书
    <rp>
      (
    </rp>
    <rt>
      shū
    </rt>
    <rp>
      )
    </rp>
  </ruby>,
  "，",
  "你",
  "呢",
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with one character having classes but no annotation should render that character with a <span> tag 1`] = `
[
  "我",
  "爱",
  "看",
  <span
    className="non-latin logogram chinese pinyin flat"
  >
    书
  </span>,
  "，",
  "你",
  "呢",
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with one character having neither annotation nor classes should render that character with no tags 1`] = `
[
  "我",
  "爱",
  "看",
  "书",
  "，",
  "你",
  "呢",
  "？",
]
`;

exports[`<DecoratedParagraph> when receiving a paragraph with one intermediate newline should render the newline as a <br/> 1`] = `
[
  "A",
  "l",
  "p",
  "h",
  "a",
  <br />,
  "B",
  "e",
  "t",
  "a",
]
`;
